### PR TITLE
Feature/oauth self hosted

### PR DESCRIPTION
**Pull request by acoumb**
Current PR contains the setup required to setup OAuth flow through a self hosted engine for following integrations: Dynamics, Semrush, Google Search Console, Hubspot.

The setting `UseUmbracoAuthorization` will toggle the OAuth flow from the default Umbraco one to a private one.

The custom settings go under the `[Integration]OAuthSettings` and contain the following properties:
- client id
- client secret
- redirect uri
- scopes
- authorization endpoint
- token endpoint
Additional keys have been added to support integrations running on V8.

For authorization flow I am using two services, the Umbraco default one and one custom. Both implement `I[Integration]AuthorizationService` with methods to retrieve the authorization URL, the access token
and refresh token.

A factory method will be then used to load the proper authorization service.

### CODE REVIEW

File: src/Umbraco.Cms.Integrations.SEO.GoogleSearchConsole.UrlInspectionTool/Configuration/GoogleSearchConsoleSettings.cs [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+            UseUmbracoAuthorization = 
+                bool.TryParse(appSettings[Constants.Configuration.UmbracoCmsIntegrationsSeoGoogleSearchConsoleUseUmbracoAuthorizationKey], out var key)
+                    ? key
+                    : true;
+        }
+
+        public string InspectUrl { get; set; }
+
+        public bool UseUmbracoAuthorization { get; set; }
```
Comment (AndyButland) [inline src/Umbraco.Cms.Integrations.SEO.GoogleSearchConsole.UrlInspectionTool/Configuration/GoogleSearchConsoleSettings.cs:23]:
```suggestion
        public bool UseUmbracoAuthorization { get; set; } = true;
```

File: src/Umbraco.Cms.Integrations.SEO.Semrush/Configuration/SemrushSettings.cs [hunk 2]
```diff
@@ -1,29 +1,23 @@
...
-        public static string SemrushAuthorizationEndpoint
-            = $"{SemrushBaseAddress}oauth2/authorize?ref=0053752252&client_id=umbraco&redirect_uri=%2Foauth2%2Fumbraco%2Fsuccess&response_type=code&scope=user.id,domains.info,url.info,positiontracking.info";
-
-        public static string SemrushKeywordsEndpoint =
-            SemrushBaseAddress + "api/v1/keywords/{0}?access_token={1}&phrase={2}&database={3}";
-
-        public static string TokenDbKey = "Umbraco.Cms.Integrations.Semrush.TokenDbKey";
-
-        public static string AllowLimitOffsetHeaderName = "Allow-Limit-Offset";
+        public bool UseUmbracoAuthorization { get; set; }
```
Comment (AndyButland) [inline src/Umbraco.Cms.Integrations.SEO.Semrush/Configuration/SemrushSettings.cs:21]:
```suggestion
        public bool UseUmbracoAuthorization { get; set; } = true;
```

File: src/Umbraco.Cms.Integrations.Crm.Dynamics/Configuration/DynamicsSettings.cs [hunk 3]
```diff
@@ -8,13 +8,19 @@ public class DynamicsSettings
...
+            UseUmbracoAuthorization = bool.TryParse(appSettings[Constants.Configuration.UmbracoCmsIntegrationsCrmDynamicsUseUmbracoAuthorizationKey], out var key) 
+                ? key 
+                : true;
         }
 
         public string HostUrl { get; set; }
 
         public string ApiPath { get; set; }
+
+        public bool UseUmbracoAuthorization { get; set; }
```
Comment (AndyButland) [inline src/Umbraco.Cms.Integrations.Crm.Dynamics/Configuration/DynamicsSettings.cs:24]:
```suggestion
        public bool UseUmbracoAuthorization { get; set; } = true;
```

**Review by AndyButland — APPROVED** [review]
Looks great, I think this will be a nice option for users to have.  Other than a few minor inline comments, the only thing missing that I can see before you merge is to add some clear documentation of the two options on the readme pages for each integration.

Maybe you could do a quick demo in the refining session we've got later too?

File: src/Umbraco.Cms.Integrations.Crm.Dynamics/DynamicsComposer.cs [hunk 4]
```diff
@@ -17,15 +17,28 @@ namespace Umbraco.Cms.Integrations.Crm.Dynamics
...
         {
             var options = builder.Services.AddOptions<DynamicsSettings>()
                 .Bind(builder.Config.GetSection(Constants.Configuration.Settings));
+            var oauthOptions = builder.Services.AddOptions<DynamicsOAuthSettings>()
+                .Bind(builder.Config.GetSection(Constants.Configuration.OAuthSettings));
 
             builder.AddNotificationHandler<UmbracoApplicationStartingNotification, UmbracoAppStartingHandler>();
 
-            builder.Services.AddSingleton<IAuthorizationService, AuthorizationService>();
+            builder.Services.AddTransient<UmbracoAuthorizationService>();
```
[the remark this answers is not in the document]
Comment (acoumb) [reply src/Umbraco.Cms.Integrations.Crm.Dynamics/DynamicsComposer.cs]: Indeed it doesn't make sense to serve a new instance every time, works fine as `Singleton`.

**Comment by acoumb** [comment]
Feedback addressed, I am leaving this open for now so you can check also on my latest comment on the OAuth Proxy update [here](https://github.com/umbraco/Umbraco.Forms.Integrations/pull/70#issuecomment-1483756848).

**Comment by acoumb** [comment]
- Included the updates for _Shopify_. 
- Removed the `AuthorizationEndpoint` property, as the service's authorization endpoint is added to the `BaseAuthorizationService`.
- Corrected the authorization controllers for Umbraco 8 to use `System.Web.Http` for the API calls on redirects.

**Review by AndyButland — APPROVED** [review]

**Pull request merged**